src/trading/Trade.sol

```diff
@@ -0,0 +1,78 @@
+// Copyright (C) 2015 Forecast Foundation OU, full GPL notice in LICENSE
+
+pragma solidity ^0.4.13;
+
+import 'ROOT/Controlled.sol';
+import 'ROOT/libraries/ReentrancyGuard.sol';
+import 'ROOT/reporting/IMarket.sol';
+import 'ROOT/trading/IOrders.sol';
+import 'ROOT/trading/Trading.sol';
+
+
+// FIXME: Remove this once MakeOrder is implemented in Solidity
+contract IMakeOrder {
+    function makeOrder(address, Trading.TradeDirections, uint256, int256, IMarket, uint8, bytes32, bytes32, uint256) returns (bytes32);
+}
+
+
+// FIXME: Remove this once TakeOrder is implemented in Solidity
+contract ITakeOrder {
+    function takeOrder(address, bytes32, Trading.TradeTypes, IMarket, uint8, uint256, uint256) returns (uint256);
+}
+
+
+contract Trade is Controlled, ReentrancyGuard {
+    uint256 private constant MINIMUM_GAS_NEEDED = 300000;
+
+    function publicBuy(IMarket _market, uint8 _outcome, uint256 _fxpAmount, int256 _fxpPrice, uint256 _tradeGroupID) external onlyInGoodTimes nonReentrant returns (bytes32) {
+        bytes32 _output = this.trade(msg.sender, Trading.TradeDirections.Buying, _market, _outcome, _fxpAmount, _fxpPrice, _tradeGroupID);
+        return _output;
+    }
+
+    function publicSell(IMarket _market, uint8 _outcome, uint256 _fxpAmount, int256 _fxpPrice, uint256 _tradeGroupID) external onlyInGoodTimes nonReentrant returns (bytes32) {
+        bytes32 _output = this.trade(msg.sender, Trading.TradeDirections.Selling, _market, _outcome, _fxpAmount, _fxpPrice, _tradeGroupID);
+        return _output;
+    }
+
+    function publicTrade(Trading.TradeDirections _direction, IMarket _market, uint8 _outcome, uint256 _fxpAmount, int256 _fxpPrice, uint256 _tradeGroupID) external onlyInGoodTimes nonReentrant returns (bytes32) {
+        bytes32 _output = this.trade(msg.sender, _direction, _market, _outcome, _fxpAmount, _fxpPrice, _tradeGroupID);
+        return _output;
+    }
+
+    function publicTakeBestOrder(Trading.TradeDirections _direction, IMarket _market, uint8 _outcome, uint256 _fxpAmount, int256 _fxpPrice, uint256 _tradeGroupID) external onlyInGoodTimes nonReentrant returns (uint256) {
+        uint256 _fxpAmountRemaining = this.takeBestOrder(msg.sender, _direction, _market, _outcome, _fxpAmount, _fxpPrice, _tradeGroupID);
+        return _fxpAmountRemaining;
+    }
+
+    function trade(address _sender, Trading.TradeDirections _direction, IMarket _market, uint8 _outcome, uint256 _fxpAmount, int256 _fxpPrice, uint256 _tradeGroupID) external onlyWhitelistedCallers returns (bytes32) {
```

Comment:
Can you check to see if `trade` is actually called from elsewhere in the contracts?  If not, then perhaps we can just mark it as `private` and not deal with the `this.` calling convention.

I suspect it _is_ called from elsewhere in the contracts so we have to continue to follow this pattern, but would be nice to verify that.

Reply:
Looks like it is only called internally so I will make it so!